### PR TITLE
Add custom headers to JDBC driver

### PR DESCRIPTION
**Pull request by basarhamdionat**
Currently, there is no way to inject custom header to JDBC driver (using REST API). This will enable JDBC driver to take customHeaders field with Key/Value mappings. Keys will be header fields and Values will be value mapping of corresponding header.

Test plan - (Please fill in how you tested your changes)
Added unit test to check (with edge cases)

Please make sure your submission complies with our [Development](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#development), [Formatting](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#formatting), and [Commit Message](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#commit-formatting-and-pull-requests) guidelines. Don't forget to follow our [attribution guidelines](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#attribution) for any code copied from other projects.

Fill in the release notes towards the bottom of the PR description.
See [Release Notes Guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines) for details.

== NO RELEASE NOTE ==

### CODE REVIEW

File: presto-client/src/main/java/com/facebook/presto/client/ClientSession.java [hunk 1]
```diff
@@ -134,6 +137,14 @@ public ClientSession(
             checkArgument(charsetEncoder.canEncode(entry.getKey()), "Credential name is not US_ASCII: %s", entry.getKey());
             checkArgument(charsetEncoder.canEncode(entry.getValue()), "Credential value is not US_ASCII: %s", entry.getValue());
         }
+
+        // verify the extra credentials are valid
+        for (Entry<String, String> entry : customHeaders.entrySet()) {
+            checkArgument(!entry.getKey().isEmpty(), "Credential name is empty");
+            checkArgument(entry.getKey().indexOf('=') < 0, "Credential name must not contain '=': %s", entry.getKey());
+            checkArgument(charsetEncoder.canEncode(entry.getKey()), "Credential name is not US_ASCII: %s", entry.getKey());
+            checkArgument(charsetEncoder.canEncode(entry.getValue()), "Credential value is not US_ASCII: %s", entry.getValue());
```
Comment (ajaygeorge) [inline presto-client/src/main/java/com/facebook/presto/client/ClientSession.java:146]: checkArguments will throw IllegalArgumentException
I think we will have to wrap them as Presto user errors.

File: presto-client/src/main/java/com/facebook/presto/client/ClientSession.java [hunk 2]
```diff
@@ -103,6 +105,7 @@ public ClientSession(
         this.preparedStatements = ImmutableMap.copyOf(requireNonNull(preparedStatements, "preparedStatements is null"));
         this.roles = ImmutableMap.copyOf(requireNonNull(roles, "roles is null"));
         this.extraCredentials = ImmutableMap.copyOf(requireNonNull(extraCredentials, "extraCredentials is null"));
+        this.customHeaders = ImmutableMap.copyOf(requireNonNull(customHeaders, "customHeaders is null"));
```
Comment (ajaygeorge) [inline presto-client/src/main/java/com/facebook/presto/client/ClientSession.java:108]: do we need a customHeader passed in mandatorily.?
Would an Optional suffice instead?

File: presto-client/src/main/java/com/facebook/presto/client/ClientSession.java [hunk 3]
```diff
@@ -134,6 +137,14 @@ public ClientSession(
             checkArgument(charsetEncoder.canEncode(entry.getKey()), "Credential name is not US_ASCII: %s", entry.getKey());
             checkArgument(charsetEncoder.canEncode(entry.getValue()), "Credential value is not US_ASCII: %s", entry.getValue());
         }
+
+        // verify the extra credentials are valid
```
Comment (ajaygeorge) [inline presto-client/src/main/java/com/facebook/presto/client/ClientSession.java:141]: I think this is a copy paste from the prev section :) 
Please update the comments.

File: presto-client/src/main/java/com/facebook/presto/client/ClientSession.java (hunk 1 above)
Comment (ajaygeorge) [reply presto-client/src/main/java/com/facebook/presto/client/ClientSession.java:146]: These checks seem to be repeated . 
Maybe have a wrapper method called `isValid(Map<String, String> properties)` to run the same check everywhere.

On a side note, curious why the resourceEstimates doesn't need a valid value check

File: presto-client/src/main/java/com/facebook/presto/client/StatementClientV1.java [hunk 4]
```diff
@@ -207,6 +207,11 @@ private Request buildQueryRequest(ClientSession session, String query)
             builder.addHeader(PRESTO_SESSION_FUNCTION, urlEncode(entry.getKey()) + "=" + urlEncode(entry.getValue()));
         }
 
+        Map<String, String> customHeaders = session.getCustomHeaders();
+        for (Entry<String, String> entry : customHeaders.entrySet()) {
+            builder.addHeader(entry.getKey(), urlEncode(entry.getValue()));
```
Comment (ajaygeorge) [inline presto-client/src/main/java/com/facebook/presto/client/StatementClientV1.java:212]: what happens if I pass in `PRESTO_SESSION_FUNCTION` as the key ?
will it override what we set in line 207

File: presto-jdbc/src/main/java/com/facebook/presto/jdbc/AbstractConnectionProperty.java [hunk 5]
```diff
@@ -192,6 +196,29 @@ private StringMapConverter() {}
         }
     }
 
+    protected static final class CustomHeadersConverter
+            implements Converter<Map<String, String>>
+    {
+        private static final CharMatcher PRINTABLE_ASCII = CharMatcher.inRange((char) 0x21, (char) 0x7E);
```
Comment (ajaygeorge) [inline presto-jdbc/src/main/java/com/facebook/presto/jdbc/AbstractConnectionProperty.java:202]: this var doesn't seem to be used anywhere.

**Comment by ajaygeorge** [comment]
@basarhamdionat Can you please add a sample JDBC connection string in the PR description, so that it would be easy for reviewers to understand the proposed change better.

File: presto-jdbc/src/main/java/com/facebook/presto/jdbc/AbstractConnectionProperty.java [hunk 6]
```diff
@@ -192,6 +196,29 @@ private StringMapConverter() {}
...
+        private CustomHeadersConverter()
+        {
+            objectMapper.configure(JsonParser.Feature.ALLOW_SINGLE_QUOTES, true);
+        }
+
+        @Override
+        public Map<String, String> convert(String value)
+        {
+            try {
+                return objectMapper.readValue(value, HashMap.class);
```
Comment (ajaygeorge) [inline presto-jdbc/src/main/java/com/facebook/presto/jdbc/AbstractConnectionProperty.java:214]: to get rid of the Unchecked assignment compiler warning
use  `TypeReference` instead.
`return objectMapper.readValue(value, new TypeReference<HashMap<String, String>>() {});`

File: presto-jdbc/src/test/java/com/facebook/presto/jdbc/TestJdbcConnection.java [hunk 7]
```diff
@@ -247,6 +257,32 @@ public void testExtraCredentials()
...
+        String customHeaders = "{'testHeaderKey':'testHeaderValue'}";
+        String encodedCustomHeaders = URLEncoder.encode(customHeaders, StandardCharsets.UTF_8.toString());
+        Connection connection = createConnection("customHeaders=" + encodedCustomHeaders);
+        assertTrue(connection instanceof PrestoConnection);
+        PrestoConnection prestoConnection = connection.unwrap(PrestoConnection.class);
+        assertEquals(prestoConnection.getCustomHeaders(), customHeadersMap);
+    }
+
+    @Test(dataProvider = "customHeaderWithSpecialCharacter")
+    public void testCustomHeadersWithSpecialCharacters(String header)
```
Comment (ajaygeorge) [inline presto-jdbc/src/test/java/com/facebook/presto/jdbc/TestJdbcConnection.java:274]: maybe rename the arg as headerValue so that is is clear

File: presto-jdbc/src/test/java/com/facebook/presto/jdbc/TestPrestoDriverUri.java [hunk 8]
```diff
@@ -256,6 +257,18 @@ public void testUriWithExtraCredentials()
...
+    @Test
+    public void testUriWithCustomHeaders()
+            throws SQLException, UnsupportedEncodingException
+    {
+        String extraCredentials = "{'testHeaderKey':'testHeaderValue'}";
+        String encodedExtraCredentials = URLEncoder.encode(extraCredentials, StandardCharsets.UTF_8.toString());
+        System.out.println("presto://localhost:8080?customHeaders=" + encodedExtraCredentials);
+        PrestoDriverUri parameters = createDriverUri("presto://localhost:8080?customHeaders=" + encodedExtraCredentials);
+        Properties properties = parameters.getProperties();
+        assertEquals(properties.getProperty(CUSTOM_HEADERS.getKey()), extraCredentials);
```
Comment (ajaygeorge) [inline presto-jdbc/src/test/java/com/facebook/presto/jdbc/TestPrestoDriverUri.java:269]: extraCredentials has a diff meaning. Please rename this method block to use custom headers as the naming

File: presto-main/src/main/java/com/facebook/presto/SessionRepresentation.java [hunk 9]
```diff
@@ -67,6 +67,7 @@
     private final Map<String, SelectedRole> roles;
     private final Map<String, String> preparedStatements;
     private final Map<SqlFunctionId, SqlInvokedFunction> sessionFunctions;
+    private final Map<String, String> customHeaders;
```
Comment (ajaygeorge) [inline presto-main/src/main/java/com/facebook/presto/SessionRepresentation.java:70]: Can you help me understand why we need changes to this class.
Are we burning in the passed in headers into the session. Is that required ?

Comment (basarhamdionat) [reply presto-main/src/main/java/com/facebook/presto/SessionRepresentation.java:70]: I did it as in AbstractTestingPrestoClient 
ClientSession is created from Session class fields which is affecting the SessionRepresentation. 
I can leave it ImmutableMap.empty() if you think is better?

File: presto-jdbc/src/test/java/com/facebook/presto/jdbc/TestJdbcConnection.java (hunk 7 above)
Comment (basarhamdionat) [reply presto-jdbc/src/test/java/com/facebook/presto/jdbc/TestJdbcConnection.java]: ✅

File: presto-jdbc/src/main/java/com/facebook/presto/jdbc/AbstractConnectionProperty.java (hunk 6 above)
Comment (basarhamdionat) [reply presto-jdbc/src/main/java/com/facebook/presto/jdbc/AbstractConnectionProperty.java]: ✅

File: presto-jdbc/src/main/java/com/facebook/presto/jdbc/AbstractConnectionProperty.java (hunk 5 above)
Comment (basarhamdionat) [reply presto-jdbc/src/main/java/com/facebook/presto/jdbc/AbstractConnectionProperty.java]: Removed it.

File: presto-client/src/main/java/com/facebook/presto/client/StatementClientV1.java (hunk 4 above)
Comment (basarhamdionat) [reply presto-client/src/main/java/com/facebook/presto/client/StatementClientV1.java:212]: I could not find any method to access the header. I am going to move it at the top so it won't override.

File: presto-jdbc/src/test/java/com/facebook/presto/jdbc/TestPrestoDriverUri.java (hunk 8 above)
Comment (basarhamdionat) [reply presto-jdbc/src/test/java/com/facebook/presto/jdbc/TestPrestoDriverUri.java]: ✅

File: presto-client/src/main/java/com/facebook/presto/client/ClientSession.java (hunk 2 above)
Comment (basarhamdionat) [reply presto-client/src/main/java/com/facebook/presto/client/ClientSession.java:108]: We can do the optional, but it seems there is no trend here
What I mean is that extraCredentials can be empty as well, but it is not optional. 
Will it impact the coding style of Presto?

Comment (ajaygeorge) [reply presto-client/src/main/java/com/facebook/presto/client/ClientSession.java:108]: lets keep it as Map for now in accordance with the other properties we are setting.

File: presto-main/src/main/java/com/facebook/presto/SessionRepresentation.java (hunk 9 above)
Comment (ajaygeorge) [reply presto-main/src/main/java/com/facebook/presto/SessionRepresentation.java:70]: yup.. lets not change Session for this. cc @tdcmeehan

Comment (basarhamdionat) [reply presto-main/src/main/java/com/facebook/presto/SessionRepresentation.java]: Removed it

**Review by ajaygeorge — APPROVED** [review]
LGTM. 
Please run a verifier as discussed to ensure that PRISM_REGION_NOT_ALLOWED are not coming up again

File: presto-jdbc/src/main/java/com/facebook/presto/jdbc/AbstractConnectionProperty.java [hunk 10]
```diff
@@ -192,6 +197,28 @@ private StringMapConverter() {}
         }
     }
 
+    protected static final class CustomHeadersConverter
```
Comment (tdcmeehan) [inline presto-jdbc/src/main/java/com/facebook/presto/jdbc/AbstractConnectionProperty.java:200]: Can we just use the `StringMapConverter` above instead?

**Review by tdcmeehan — CHANGES_REQUESTED** [review]
Please also update the JDBC client documentation: https://github.com/prestodb/presto/blob/master/presto-docs/src/main/sphinx/installation/jdbc.rst

File: presto-jdbc/src/main/java/com/facebook/presto/jdbc/AbstractConnectionProperty.java (hunk 10 above)
Comment (basarhamdionat) [reply presto-jdbc/src/main/java/com/facebook/presto/jdbc/AbstractConnectionProperty.java:200]: No StringMapConverter splits by colon and it could impact the parsing if some custom header with hostport is passed.

**Comment by basarhamdionat** [comment]
> Please also update the JDBC client documentation: https://github.com/prestodb/presto/blob/master/presto-docs/src/main/sphinx/installation/jdbc.rst

Made changes to rst please take a look.

File: presto-jdbc/src/main/java/com/facebook/presto/jdbc/AbstractConnectionProperty.java (hunk 10 above)
Comment (tdcmeehan) [reply presto-jdbc/src/main/java/com/facebook/presto/jdbc/AbstractConnectionProperty.java:200]: How about we percent-encode the values for this field?  That way we don't need to have two different ways to pass in a map value.

Comment (basarhamdionat) [reply presto-jdbc/src/main/java/com/facebook/presto/jdbc/AbstractConnectionProperty.java]: OK Tim I have updated, please let me know if it looks good

File: presto-docs/src/main/sphinx/installation/jdbc.rst [hunk 11]
```diff
@@ -103,4 +103,9 @@ Name                              Description
 ``extraCredentials``              Extra credentials for connecting to external services. The
                                   extraCredentials is a list of key-value pairs. Example:
                                   ``foo:bar;abc:xyz`` will create credentials ``abc=xyz`` and ``foo=bar``
+``customHeaders``                 Custom headers to inject through JDBC driver. The
+                                  customHeaders is a list of key-value pairs. Example:
+                                  ``testHeaderKey:testHeaderValue`` will inject the header ``testHeaderKey``
+                                  with value ``testHeaderValue``. Need to encode key-value fields individually
+                                  if there is ``:`` in the fields.
```
Comment (tdcmeehan) [inline presto-docs/src/main/sphinx/installation/jdbc.rst:110]:
```suggestion
                                  ``testHeaderKey:testHeaderValue`` will inject the header ``testHeaderKey``
                                  with value ``testHeaderValue``. Values should be percent encoded.
```

**Review by tdcmeehan — APPROVED** [review]
LGTM % nits

File: presto-docs/src/main/sphinx/installation/jdbc.rst (hunk 11 above)
Comment (basarhamdionat) [reply presto-docs/src/main/sphinx/installation/jdbc.rst]: ✅

File: presto-jdbc/src/test/java/com/facebook/presto/jdbc/TestJdbcConnection.java [hunk 12]
```diff
@@ -247,6 +257,32 @@ public void testExtraCredentials()
...
+        String customHeaders = "testHeaderKey:testHeaderValue";
+        String encodedCustomHeaders = URLEncoder.encode(customHeaders, StandardCharsets.UTF_8.toString());
+        Connection connection = createConnection("customHeaders=" + encodedCustomHeaders);
+        assertTrue(connection instanceof PrestoConnection);
+        PrestoConnection prestoConnection = connection.unwrap(PrestoConnection.class);
+        assertEquals(prestoConnection.getCustomHeaders(), customHeadersMap);
+    }
+
+    @Test(dataProvider = "customHeaderWithSpecialCharacter")
+    public void testCustomHeadersWithSpecialCharacters(String testHeaderKey)
```
[the remark this answers is not in the document]
Comment (basarhamdionat) [reply presto-jdbc/src/test/java/com/facebook/presto/jdbc/TestJdbcConnection.java]: ✅
